### PR TITLE
Integrate 'Extract Widget' refactoring.

### PR DESCRIPTION
**Pull request by scheglov**
image

https://github.com/flutter/flutter-intellij/issues/1250

### CODE REVIEW

File: resources/META-INF/plugin.xml [hunk 1]
```diff
@@ -192,6 +192,11 @@
       <add-to-group group-id="RunMenu" anchor="after" relative-to-action="Stop"/>
     </group>
 
+    <!-- refactoring menu -->
+    <action class="io.flutter.actions.ExtractWidgetAction" id="Flutter.ExtractWidget" text="Flutter Widget...">
```
Comment (devoncarew) [inline resources/META-INF/plugin.xml]: Will this show up in an 'extract' menu? I'm wondering if the user will have enough context to know that `Flutter Widget...` means extract flutter widget. (if so, then lets keep the shorter text)

File: src/io/flutter/actions/ExtractWidgetAction.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+  public ExtractWidgetDialog(@NotNull Project project,
+                             @Nullable Editor editor,
+                             @NotNull ExtractWidgetRefactoring refactoring) {
+    super(project, editor, refactoring);
+    myRefactoring = refactoring;
+    setTitle("Extract Widget");
+    init();
+
+    myNameField.setText("NewWidget");
```
Comment (devoncarew) [inline src/io/flutter/actions/ExtractWidgetAction.java]: We don't know whether this new name will be legal `NewWidget` - if we did we could do something like `NewWidget7`.

Another option is to autogenerate interesting names that users would need to change. So, `BluePony` or `TinySunrise`; this is what DartPad does. I'm not advocating for this, but thought I would mention it.

File: resources/META-INF/plugin.xml (hunk 1 above)
Comment (scheglov) [reply resources/META-INF/plugin.xml]: The name and location is consistent with other `extract` refactorings.

image

**Comment by slightfoot** [comment]
@scheglov @devoncarew Is it possible to get this option in the Quick Assist popup menu?
Re: https://twitter.com/devangelslondon/status/988779563990814720

**Comment by scheglov** [comment]
This is not something that can be easily done for IntelliJ.

Quick Assists are contributed by the Dart plugin, and they show only the specific set of actions returned from Dart Analysis Server. There are no ways to include arbitrary actions, from the Flutter plugin.

But you can add a keyboard shortcut, so it will be easily accessible.
image
image

**Comment by slightfoot** [comment]
I still think its worth looking into as an addition, the quick actions is the go to solution for Widget actions when doing Flutter development.

**Comment by dahabit** [comment]
@slightfoot Thanks for your suggestion, @scheglov it will be great if we can implement that

**Pull request merged**